### PR TITLE
(PUP-9121) make puppet apply with --catalog resolve deferred

### PR DESCRIPTION
**Pull request by hlindberg**
This adds resolution of deferred values to `puppet apply --catalog <file>` (and when file is stdin).
As part this, the existing logic was DRYed up or too much logic would have been duplicated between the `main` and `apply/read_catalog` methods.

### CODE REVIEW

File: lib/puppet/application/apply.rb [hunk 1]
```diff
@@ -350,18 +310,98 @@ def setup
...
+      # on (seemingly the right thing to do)), but that would remove the ability to test what happens when a
+      # rich catalog is processed without rich_data being turned on.
+      format = Puppet::Resource::Catalog.default_format
+      begin
+        catalog = Puppet::Resource::Catalog.convert_from(format, text)
+      rescue => detail
+        raise Puppet::Error, _("Could not deserialize catalog from %{format}: %{detail}") % { format: format, detail: detail }, detail.backtrace
+      end
+      # Resolve all deferred values and replace them / mutate the catalog
+      Puppet::Pops::Evaluator::DeferredResolver.resolve_and_replace(node, node.facts, catalog)
```
Comment (tvpartytonight) [inline lib/puppet/application/apply.rb]: The signature for this method was changes to allow for cached_catalogs to work in #7022, so it should just be `(node.facts, catalog)`.

**Comment by tvpartytonight** [comment]
Is this actually gated to work with only `--catalog`? I may have missed it, but it seems like this would work in any apply situation.

File: lib/puppet/application/apply.rb (hunk 1 above)
Comment (hlindberg) [reply lib/puppet/application/apply.rb]: fixed

**Comment by hlindberg** [comment]
@tvpartytonight `puppet apply` has two different code paths depending on if the command is with our without `--catalog`. The logic on the `--catalog` path did not include resolution of deferred values; that is what this PR is about. The difference is quite big between the two as without `--catalog` apply will use the catalog as it was produced by the compiler (i.e. no serialization/deserialization takes place). While there are common sequences of logic, I only took it part of the way to a full refactoring - mostly to avoid the risk of inadvertent introduction of bugs / incompatibilities - I felt the small refactoring; replacing 2-3 lines with a function call was safe enough.

**Comment by hlindberg** [comment]
AppVeyor barfed on something unrelated - rerunning.

**Comment by tvpartytonight** [comment]
jenkins please test this

**Pull request merged**